src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -66,21 +75,58 @@ public async Task<BundleResponse> Handle(BundleRequest bundleRequest, Cancellati
         {
             var bundleResource = bundleRequest.Bundle.ToPoco<Hl7.Fhir.Model.Bundle>();
 
-            if (bundleResource.Type != Hl7.Fhir.Model.Bundle.BundleType.Batch)
+            await FillRequestLists(bundleResource.Entry);
+
+            if (bundleResource.Type == Hl7.Fhir.Model.Bundle.BundleType.Batch)
             {
-                throw new MethodNotAllowedException(Microsoft.Health.Fhir.Api.Resources.OnlyCertainBundleTypesSupported);
+                var responseBundle = new Hl7.Fhir.Model.Bundle
+                {
+                    Type = Hl7.Fhir.Model.Bundle.BundleType.BatchResponse,
+                };
+
+                await ExecuteAllRequests(responseBundle);
+                return new BundleResponse(responseBundle.ToResourceElement());
             }
 
-            await FillRequestLists(bundleResource.Entry);
+            if (bundleResource.Type == Hl7.Fhir.Model.Bundle.BundleType.Transaction)
+            {
+                var responseBundle = new Hl7.Fhir.Model.Bundle
+                {
+                    Type = Hl7.Fhir.Model.Bundle.BundleType.TransactionResponse,
+                };
+
+                return await ExecuteTransactionForAllRequests(responseBundle);
```

Comment:
CosmosDB doesn't support trasaction. Instead of failing with 500 (since the CosmosFhirDataStore would throw NotImplementedException), should this method check against some sort of "capability" and returned not supported exception if the operation is not supported?

Reply:
Yes,  the same behavior happens here. Execution will reach to bundlehandler after validating the capabilitystatement. For cosmosDB, transaction will be turned off and Not Supported exception will be thrown.